### ISSUE TITLE
Implement lexicase selection

### ISSUE THREAD

**Issue by rhiever** [rhiever/edd#6]
Currently, we subset the MNIST training set down to 100 instances of each class. This is because it takes too long to train on the entire training set. However, it may be better to use lexicase selection, which will likely require fewer evaluations per generation.

It works as follows:

1) Each round, randomize the order of the MNIST training set

2) Evaluate all remaining classifiers on the next training instance

3) "Eliminate" all classifiers that failed to classify `2)` correctly -- they lose their chance to reproduce in this round

4) Repeat steps `2)`-`3)` until only one classifier remains in contention. If all remaining classifiers get eliminated at once, choose one classifier randomly from the remaining classifiers in the last step of the round.

5) The last remaining classifier asexually reproduces one offspring into the next generation.

6) Repeat steps `1)` through `5)` until the new population is fully created.

**Comment by Vaguery** [comment]
You may be able to save a lot of effort in a large dataset by doing this lazily.

**Comment by mercere99** [comment]
Also, regarding steps 3-4, if everyone gets eliminated in a single step, you may want to keep them all and just keep testing. Stop when you've tested one from each category. I think this will be especially important early on when they won't be very good at all and you'll mostly be picking randomly from the whole population if you don't keep going.

**Comment by rhiever** [comment]
@Vaguery: Of course! It will not be feasible to do it with exhaustive evaluation every generation, as the optimization process is quite slow even with just a data set of 1,000 training instances.

@mercere99: 👍

**Comment by Vaguery** [comment]
Actually, technically you've got the algorithm wrong; I just realized based on @mercere99's comment. To select one individual:
1. permute the training set
2. reduce the population by removing all except the _best-scoring_ individuals on the first training case; if everybody has the same score in on that case, they're all "best-scoring"
3. repeat 2 for each training case
4. if multiple individuals remain, pick one uniformly

The Spector crowd always use "positive selection", as opposed to biologically-analogous selection. That is, they use "selection" to mean "picking who to keep", not the more biologically proper "biasing an otherwise uniform death process".
